core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,428 @@
+          } else {
+            n = readFromUnderlyingStream(inBuffer)
+          }
+        }
+        if (n <= 0) {
+          n
+        } else {
+          streamOffset += n // Read n bytes
+          decrypt(decryptor, inBuffer, outBuffer, padding)
+          padding = afterDecryption(decryptor, inBuffer, streamOffset, iv)
+          n = Math.min(len, outBuffer.remaining())
+          outBuffer.get(b, off, n)
+          n
+        }
+      }
+    }
+  }
+
+  /** Read data from underlying stream. */
+  def readFromUnderlyingStream(inBuffer: ByteBuffer): Int = {
+    val toRead: Int = inBuffer.remaining
+    val tmp: Array[Byte] = getTmpBuf
+    val n: Int = in.read(tmp, 0, toRead)
+    if (n > 0) {
+      inBuffer.put(tmp, 0, n)
+    }
+    n
+  }
+
+
+  def getTmpBuf: Array[Byte] = {
+    if (tmpBuf == null) {
+      tmpBuf = new Array[Byte](bufferSize)
+    }
+    tmpBuf
+  }
+
+  /**
+   * Do the decryption using inBuffer as input and outBuffer as output.
+   * Upon return, inBuffer is cleared the decrypted data starts at
+   * outBuffer.position() and ends at outBuffer.limit()
+   */
+  def decrypt(decryptor: Decryptor, inBuffer: ByteBuffer, outBuffer: ByteBuffer, padding: Byte) {
+    Preconditions.checkState(inBuffer.position >= padding)
+    if (inBuffer.position != padding) {
+      inBuffer.flip
+      outBuffer.clear
+      decryptor.decrypt(inBuffer, outBuffer)
+      inBuffer.clear
+      outBuffer.flip
+      if (padding > 0) {
+        outBuffer.position(padding)
+      }
+    }
+  }
+
+  /**
+   * This method is executed immediately after decryption. Check whether
+   * decryptor should be updated and recalculate padding if needed.
+   */
+  def afterDecryption(decryptor: Decryptor, inBuffer: ByteBuffer, position: Long,
+                      iv: Array[Byte]): Byte = {
+    var padding: Byte = 0
+    if (decryptor.isContextReset) {
+      updateDecryptor(decryptor, position, iv)
+      padding = getPadding(position)
+      inBuffer.position(padding)
+    }
+    padding
+  }
+
+  def getCounter(position: Long): Long = {
+    position / codec.getCipherSuite.algoBlockSize
+  }
+
+  def getPadding(position: Long): Byte = {
+    (position % codec.getCipherSuite.algoBlockSize).asInstanceOf[Byte]
+  }
+
+  /** Calculate the counter and iv, update the decryptor. */
+  def updateDecryptor(decryptor: Decryptor, position: Long, iv: Array[Byte]) {
+    val counter: Long = getCounter(position)
+    codec.calculateIV(initIV, counter, iv)
+    decryptor.init(key, iv)
+  }
+
+  /**
+   * Reset the underlying stream offset clear {@link #inBuffer} and
+   * {@link #outBuffer}. This Typically happens during {@link #seek(long)}
+   * or {@link #skip(long)}.
+   */
+  def resetStreamOffset(offset: Long) {
+    streamOffset = offset
+    inBuffer.clear
+    outBuffer.clear
+    outBuffer.limit(0)
+    updateDecryptor(decryptor, offset, iv)
+    padding = getPadding(offset)
+    inBuffer.position(padding)
+  }
+
+  override def close {
```

Comment:
see [https://issues.apache.org/jira/browse/HADOOP-11710](https://issues.apache.org/jira/browse/HADOOP-11710) for sync logic here & elsewhere